docs/internals/optimizer.rst

```diff
@@ -65,31 +65,37 @@ A "runs" parameter of "1" will produce short but expensive code. The largest val
 Opcode-Based Optimizer Module
 =============================
 
-The opcode-based optimizer module operates on assembly. It splits the
+The opcode-based optimizer module operates on assembly code. It splits the
 sequence of instructions into basic blocks at ``JUMPs`` and ``JUMPDESTs``.
 Inside these blocks, the optimizer analyzes the instructions and records every modification to the stack,
 memory, or storage as an expression which consists of an instruction and
-a list of arguments which are pointers to other expressions. The opcode-based optimizer
-uses a component called "CommonSubexpressionEliminator" that amongst other
+a list of arguments which are pointers to other expressions. 
+
+Additionally, the opcode-based optimizer
+uses a component called "CommonSubexpressionEliminator" that, amongst other
 tasks, finds expressions that are always equal (on every input) and combines
 them into an expression class. It first tries to find each new
 expression in a list of already known expressions. If no such matches are found,
 it simplifies the expression according to rules like
 ``constant + constant = sum_of_constants`` or ``X * 1 = X``. Since this is
 a recursive process, we can also apply the latter rule if the second factor
-is a more complex expression where we know that it always evaluates to one.
-Modifications to storage and memory locations have to erase knowledge about
-storage and memory locations which are not known to be different. If we first
-write to location x and then to location y and both are input variables, the
-second could overwrite the first, so we do not know what is stored at x after
-we wrote to y. If simplification of the expression ``x - y`` evaluates to a
-non-zero constant, we know that we can keep our knowledge about what is stored at ``x``.
+is a more complex expression which we know always evaluates to one.
+
+Certain optimizer steps symbolically tracks the storage and memory locations. Therefore
+modifications to storage and memory locations, of say location ``l``, must erase knowledge about
+storage or memory locations which are not known to be different from ``l``. For example, If we first
+write to a symbolic storage location ``x`` and then to the symbolic storage location ``y``. In
+general, ``y`` could overwrite ``x`` if ``y`` and ``x`` are the same storage slot. However, if the
+optimizer can infer that the value of the expression ``x - y`` is non-zero, this means that the
+knowledge about storage at the symbolic location ``x`` can be kept.
```

Comment:
Updated this. @leonardoalt Perhaps a review of at least this part?

Reply:
Hmm I don't understand how it arrives at the implication that knowledge can be kept if the expression results in non-zero...